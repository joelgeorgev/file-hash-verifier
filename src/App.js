import React, { Component } from 'react';
import FilePicker from './components/file-picker';
import HashSelector from './components/hash-selector';
import FileLoader from './components/file-loader';
import FileDetails from './components/file-details';
import FileHash from './components/file-hash';
import HashVerifier from './components/hash-verifier';
import './App.css';

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      file: undefined,
      fileLoadStatus: 0,
      arrayBuffer: undefined,
      hashType: 'sha-1',
      hash: undefined
    }
  }

  async setFile(e) {
    const files = e.dataTransfer && e.dataTransfer.files ? e.dataTransfer.files : e.target.files;
    if (files.length) {
      this.setState({
        file: files[0],
        fileLoadStatus: 0,
        arrayBuffer: undefined,
        hash: ''
      });
      try {
        const arrayBuffer = await this.getArrayBuffer(files[0]);
        const hash = await this.hash(this.state.hashType, arrayBuffer);
        this.setState({
          arrayBuffer: arrayBuffer,
          hash: hash
        });
      } catch (err) {
        console.error('Error', err);
      }
    }
  }

  async setHashType(e) {
    const hashType = e.target.value;
    this.setState({
      hashType: hashType,
      hash: ''
    });
    if (this.state.arrayBuffer) {
      try {
        const hash = await this.hash(hashType, this.state.arrayBuffer);
        this.setState({
          hash: hash
        });
      } catch (err) {
        console.error('Error', err);
      }
    }
  }

  async hash(hashType, arrayBuffer) {
    const hashBuffer = await crypto.subtle.digest(hashType, arrayBuffer);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    const hashHex = hashArray.map(b => ('00' + b.toString(16)).slice(-2)).join('');
    return hashHex;
  }

  getArrayBuffer(file) {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => { resolve(reader.result); };
      reader.onprogress = e => { this.setState({ fileLoadStatus: Math.round((e.loaded / e.total) * 100) }); };
      reader.onerror = err => { reject(err); }
      reader.readAsArrayBuffer(file);
    });
  }

  render() {
    return (
      <div className='flex flex-column justify-center ma4'>
        <FilePicker fileLoadStatus={this.state.fileLoadStatus} setFile={this.setFile.bind(this)} />
        <HashSelector hashType={this.state.hashType} setHashType={this.setHashType.bind(this)} />
        {this.state.file ?
          <div>
            <FileLoader fileLoadStatus={this.state.fileLoadStatus} />
            <FileDetails file={this.state.file} />
            <FileHash fileLoadStatus={this.state.fileLoadStatus} hash={this.state.hash} />
            <HashVerifier hash={this.state.hash} />
          </div>
          :
          ''}
      </div>
    );
  }
}

export default App;