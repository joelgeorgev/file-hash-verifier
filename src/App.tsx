import { createStore, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'
import createSagaMiddleware from 'redux-saga'
import styled from 'styled-components'

import { ShellContainer } from './containers'
import { reducer } from './reducers'
import { saga } from './sagas'
import github from './assets/github.svg'

const sagaMiddleware = createSagaMiddleware()
const store = createStore(reducer, applyMiddleware(sagaMiddleware))
sagaMiddleware.run(saga)

const Main = styled.main`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 80%;
  max-width: 64rem;
  height: 100vh;
  margin: 0 auto;
`

const Header = styled.header`
  text-align: center;
`

const Section = styled.section`
  display: flex;
  flex-direction: column;
  flex: 1;
  overflow: auto;
`

const Footer = styled.footer`
  margin: 2rem;
`

export const App = () => (
  <Main>
    <Header>
      <h1>File Hash Verifier</h1>
    </Header>
    <Section>
      <article>
        <Provider store={store}>
          <ShellContainer />
        </Provider>
      </article>
    </Section>
    <Footer>
      <a href='https://github.com/joelgeorgev/file-hash-verifier'>
        <img src={github} alt='Go to GitHub repository page' />
      </a>
    </Footer>
  </Main>
)
