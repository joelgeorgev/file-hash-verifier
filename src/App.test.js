import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import { App } from './App';

it('smoke test', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});

it('snapshot test', () => {
  const tree = renderer.create(
    <App />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('radio button state change test', () => {
  let app = ReactTestUtils.renderIntoDocument(<App />);

  // Get 'SHA-256' radio button node
  let inputElement = ReactTestUtils.scryRenderedDOMComponentsWithTag(app, 'input')[2];
  ReactTestUtils.Simulate.change(inputElement);
  expect(inputElement.checked).toBe(true);
  expect(app.state.hashType).toBe('sha-256');
});