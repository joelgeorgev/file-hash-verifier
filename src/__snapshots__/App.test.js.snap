// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="flex flex-column justify-center ma4"
>
  <div
    className="flex flex-column justify-center h4 ba b--dashed b--light-silver"
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <label
      className="self-center pointer b"
      htmlFor="file-picker"
    >
      Drop your file here or click to pick one.
    </label>
    <input
      className="dn"
      id="file-picker"
      multiple={false}
      onChange={[Function]}
      type="file"
    />
  </div>
  <div
    className="flex justify-center mt4 b"
  >
    <input
      checked={true}
      className="mh2"
      name="hash-type"
      onChange={[Function]}
      type="radio"
      value="sha-1"
    />
    SHA-1
    <input
      checked={false}
      className="mh2"
      name="hash-type"
      onChange={[Function]}
      type="radio"
      value="sha-256"
    />
    SHA-256
    <input
      checked={false}
      className="mh2"
      name="hash-type"
      onChange={[Function]}
      type="radio"
      value="sha-384"
    />
    SHA-384
    <input
      checked={false}
      className="mh2"
      name="hash-type"
      onChange={[Function]}
      type="radio"
      value="sha-512"
    />
    SHA-512
  </div>
  
</div>
`;
